scanpipe/pipes/symbolmap.py

```diff
@@ -0,0 +1,156 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# http://nexb.com and https://github.com/aboutcode-org/scancode.io
+# The ScanCode.io software is licensed under the Apache License version 2.0.
+# Data generated with ScanCode.io is provided as-is without warranties.
+# ScanCode is a trademark of nexB Inc.
+#
+# You may not use this software except in compliance with the License.
+# You may obtain a copy of the License at: http://apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# Data Generated with ScanCode.io is provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+# ScanCode.io should be considered or used as legal advice. Consult an Attorney
+# for any legal advice.
+#
+# ScanCode.io is a free software code scanning tool from nexB Inc. and others.
+# Visit https://github.com/aboutcode-org/scancode.io for support and download.
+
+from collections import Counter
+
+from aboutcode.pipeline import LoopProgress
+from scanpipe.models import CodebaseRelation
+from scanpipe.pipes import flag
+
+"""
+Path matching using source and binary symbols.
+
+The approach is to create a set of symbols obtained from the rust binary for
+each of them and match them to the symbols obtained from the source
+"""
+
+
+def map_resources_with_symbols(
+    to_resource, from_resources, binary_symbols, map_type, logger=None
+):
+    """
+    Map paths found in the ``to_resource`` extra_data to paths of the ``from_resources``
+    CodebaseResource queryset using the precomputed ``from_resources_index`` path index.
+    """
+    if not binary_symbols:
+        return
+
+    # Accumulate unique relation objects for bulk creation
+    relations_to_create = {}
+
+    # These are of type string
+    paths_not_mapped = to_resource.extra_data[f"{map_type}_not_mapped"] = []
+    for item in match_source_paths_to_binary(
+        to_resource=to_resource,
+        from_resources=from_resources,
+        binary_symbols=binary_symbols,
+        map_type=map_type,
+        logger=logger,
+    ):
+        if isinstance(item, str):
+            paths_not_mapped.append(item)
+        else:
+            rel_key, relation = item
+            if rel_key not in relations_to_create:
+                relations_to_create[rel_key] = relation
+
+    # If there are any non-test files in the rust source files which
+    # are not mapped, we mark the binary as REQUIRES_REVIEW
+    if paths_not_mapped and any(
+        [True for path in paths_not_mapped if "/tests/" not in path]
+    ):
+        to_resource.status = flag.REQUIRES_REVIEW
+        to_resource.save()
```

Comment:
`to_resource.update(status=flag.REQUIRES_REVIEW)` would be a bit better, single line, and the `save(update_fields=)` is applied, which limit the save to a single field in that case.

Reply:
Thanks, this is much better.